### PR TITLE
Bug 1673806: build - support UTF-8 filenames in binary builds

### PR DESCRIPTION
**Pull request by adambkaplan**
* Set default LANG env var in builds for bsdtar.
* Add support for the builder's LANG env var to be overridden.

### CODE REVIEW

File: pkg/build/controller/strategy/custom.go [hunk 1]
```diff
@@ -57,7 +57,10 @@ func (bs *CustomBuildStrategy) CreateBuildPod(build *buildv1.Build, additionalCA
 		return nil, fmt.Errorf("failed to encode the build: %v", err)
 	}
 
-	containerEnv := []corev1.EnvVar{{Name: "BUILD", Value: string(data)}}
+	containerEnv := []corev1.EnvVar{
+		{Name: "BUILD", Value: string(data)},
+		{Name: "LANG", Value: "en_US.utf8"},
```
Comment (adambkaplan) [inline pkg/build/controller/strategy/custom.go]: Set the `LANG` environment variable for all containers in the build pod.

Binary builds use `bsdtar` to extract the content from STDIN. `LANG` needs to be set in order for the bsdtar to extract UTF-8 filenames.

File: pkg/build/util/consts.go [hunk 2]
```diff
@@ -3,7 +3,7 @@ package util
 // TODO: This list needs triage and move to openshift/api and library-go:
 
 var (
-	WhitelistEnvVarNames = []string{"BUILD_LOGLEVEL", "GIT_SSL_NO_VERIFY", "HTTP_PROXY", "HTTPS_PROXY", "NO_PROXY"}
+	WhitelistEnvVarNames = []string{"BUILD_LOGLEVEL", "GIT_SSL_NO_VERIFY", "HTTP_PROXY", "HTTPS_PROXY", "LANG", "NO_PROXY"}
```
Comment (adambkaplan) [inline pkg/build/util/consts.go]: Added `LANG` to the whitelist so builds can override the `en_US.utf8` default.

File: pkg/build/util/util_test.go [hunk 3]
```diff
@@ -1,171 +1,134 @@
...
-	if output[3].Value != "source" {
-		t.Errorf("Expected output env 'BUILD_LOGLEVEL' to have value 'source', got %+v", output[3])
+	testCases := []struct {
+		name                string
+		useSourcePrecedence bool
+		whitelist           []string
+		input               []corev1.EnvVar
+		currentOutput       []corev1.EnvVar
+		expectedOutput      []corev1.EnvVar
+	}{
```
Comment (adambkaplan) [inline pkg/build/util/util_test.go]: Refactored to use table-driven tests.

**Comment by adambkaplan** [comment]
@gabemontero certainly doable - I believe that there is an existing binary --from-dir test we can leverage.

**Comment by gabemontero** [comment]
cool - give it a go then @adambkaplan

File: test/extended/testdata/builds/test-build-app/utf8_tést_app.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+require 'rack'
+ 
+app = Proc.new do |env|
+    ['200', {'Content-Type' => 'text/html'}, ['Wë súpport UTF-8!']]
+end
```
Comment (adambkaplan) [inline test/extended/testdata/builds/test-build-app/utf8_tést_app.rb]: @gabemontero this dir is run in our extended `start-build` tests - see https://github.com/openshift/origin/blob/master/test/extended/builds/start.go#L172-L182

Comment (gabemontero) [reply test/extended/testdata/builds/test-build-app/utf8_tést_app.rb]: ah ... clever turn on things ... works for me :-)

File: test/extended/testdata/builds/test-build-app/utf8_tést_app.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+require 'rack'
+ 
+app = Proc.new do |env|
+    ['200', {'Content-Type' => 'text/html'}, ['Wë súpport UTF-8!']]
```
Comment (sosiouxme) [inline test/extended/testdata/builds/test-build-app/utf8_tést_app.rb]: @adambkaplan this has uncovered a bug in our downstream tools. We use tito to build the RPM from source, and apparently it's never had to handle non-ascii in a filename before. It breaks. So I'm evaluating best course of action. Fixing tito may be hard. Deleting this file is easy. If I delete this file in the private repo, am I going to run into a boatload of merge conflicts down the road? Or will this never change? Would this test be just as effective with an ascii filename? Just evaluating alternatives, please opine.

Comment (sosiouxme) [reply test/extended/testdata/builds/test-build-app/utf8_tést_app.rb]: also you should have put an emoji in there just to visually distinguish from latin-1 :)

Comment (adambkaplan) [reply test/extended/testdata/builds/test-build-app/utf8_tést_app.rb]: @sosiouxme This test won't be as effective, as the non-ascii file name verifies a bug fix in a black-box fashion. It's a trivial bug, so I'm OK deleting this and adding a separate test.

**Pull request merged**